### PR TITLE
add laplacian pyramid

### PR DESCRIPTION
**Pull request by lafith**
#### Changes
add build_laplacian_pyramid into kornia.geometry.transform.pyramid.

Fixes # (issue)

#### Type of change

- [ ] 📚  Documentation Update
- [ ]  🧪 Tests Cases
- [ ] 🔬 New feature (non-breaking change which adds functionality)

#### Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: kornia/geometry/transform/pyramid.py [hunk 1]
```diff
@@ -350,3 +350,49 @@ def build_pyramid(
         pyramid.append(img_down)
 
     return pyramid
+
+
+def build_laplacian_pyramid(
+    input: torch.Tensor, max_level: int, border_type: str = 'reflect', align_corners: bool = False
```
Comment (edgarriba) [inline kornia/geometry/transform/pyramid.py:356]:
```suggestion
    input: Tensor, max_level: int, border_type: str = 'reflect', align_corners: bool = False
```
`from korna.core import Tensor`

File: kornia/geometry/transform/pyramid.py [hunk 2]
```diff
@@ -350,3 +350,49 @@ def build_pyramid(
...
+) -> List[torch.Tensor]:
```
Comment (edgarriba) [inline kornia/geometry/transform/pyramid.py:357]:
```suggestion
) -> List[Tensor]:
```

File: kornia/geometry/transform/pyramid.py [hunk 3]
```diff
@@ -350,3 +350,49 @@ def build_pyramid(
...
+    r"""Construct the Laplacian pyramid for an image.
+
+    .. image:: _static/img/build_pyramid.png
```
Comment (edgarriba) [inline kornia/geometry/transform/pyramid.py:360]: to make this happen, you need to update this script
https://github.com/kornia/kornia/blob/master/docs/generate_example_images.py

File: kornia/geometry/transform/pyramid.py [hunk 4]
```diff
@@ -350,3 +350,49 @@ def build_pyramid(
         pyramid.append(img_down)
 
     return pyramid
+
+
+def build_laplacian_pyramid(
+    input: torch.Tensor, max_level: int, border_type: str = 'reflect', align_corners: bool = False
+) -> List[torch.Tensor]:
+    r"""Construct the Laplacian pyramid for an image.
```
Comment (edgarriba) [inline kornia/geometry/transform/pyramid.py:358]:
```suggestion
    r"""Construct the Laplacian pyramid for a tensor image.
```

File: kornia/geometry/transform/pyramid.py [hunk 5]
```diff
@@ -350,3 +350,49 @@ def build_pyramid(
...
+
+    .. image:: _static/img/build_pyramid.png
+
```
Comment (edgarriba) [inline kornia/geometry/transform/pyramid.py:361]: add the citation of this paper: https://ieeexplore.ieee.org/document/1095851

check other implementations to see how we include in bibtext

```
@ARTICLE{1095851,

  author={Burt, P. and Adelson, E.},

  journal={IEEE Transactions on Communications}, 

  title={The Laplacian Pyramid as a Compact Image Code}, 

  year={1983},

  volume={31},

  number={4},

  pages={532-540},

  doi={10.1109/TCOM.1983.1095851}}

```

File: kornia/geometry/transform/pyramid.py [hunk 6]
```diff
@@ -350,3 +350,49 @@ def build_pyramid(
...
+    The function constructs a vector of images and builds the Laplacian pyramid
+    by recursively computing the difference after applying
+    pyrUp to the adjacent layer in it's Gaussian pyramid.
+
+    Args:
+        input : the tensor to be used to construct the pyramid.
```
Comment (edgarriba) [inline kornia/geometry/transform/pyramid.py:367]:
```suggestion
        input: the tensor to be used to construct the pyramid with shape :math:`(B, C, H, W)`.
```
for free functions we tend to not use the `Shape` directive

File: kornia/geometry/transform/pyramid.py [hunk 7]
```diff
@@ -350,3 +350,49 @@ def build_pyramid(
...
+        max_level: 0-based index of the last (the smallest) pyramid layer.
+          It must be non-negative.
+        border_type: the padding mode to be applied before convolving.
+          The expected modes are: ``'constant'``, ``'reflect'``,
+          ``'replicate'`` or ``'circular'``.
+        align_corners: interpolation flag.
```
Comment (edgarriba) [inline kornia/geometry/transform/pyramid.py:373]: where is this used ?

File: kornia/geometry/transform/pyramid.py [hunk 8]
```diff
@@ -350,3 +350,49 @@ def build_pyramid(
...
+
+    Shape:
+        - Input: :math:`(B, C, H, W)`
+        - Output :math:`[(B, C, H, W), (B, C, H/2, W/2), ...]`
+    """
+    if not isinstance(input, torch.Tensor):
```
Comment (edgarriba) [inline kornia/geometry/transform/pyramid.py:379]: use the new kornia testing API for checkings: 
https://github.com/kornia/kornia/blob/master/kornia/geometry/line.py#L162

File: kornia/geometry/transform/pyramid.py (hunk 1 above)
Comment (edgarriba) [reply kornia/geometry/transform/pyramid.py:356]: same for rest

File: kornia/geometry/transform/pyramid.py [hunk 9]
```diff
@@ -350,3 +350,49 @@ def build_pyramid(
...
+    if not isinstance(max_level, int) or max_level < 0:
+        raise ValueError(f"Invalid max_level, it must be a positive integer. Got: {max_level}")
+
+    # create gaussian pyramid
+    gaussian_pyramid: List[torch.Tensor] = build_pyramid(input, max_level)
+    # create empty list
+    laplacian_pyramid: List[torch.Tensor] = []
+
+    for i in range(max_level - 1):
+        img_expand: torch.Tensor = pyrup(gaussian_pyramid[i + 1])
```
Comment (edgarriba) [inline kornia/geometry/transform/pyramid.py:394]:
```suggestion
        img_expand: torch.Tensor = pyrup(gaussian_pyramid[i + 1], align_corners)
```
@ducha-aiki not clear to me whether we should just fix the align corners here, even inside the pyrup.

File: kornia/geometry/transform/pyramid.py [hunk 10]
```diff
@@ -350,3 +350,49 @@ def build_pyramid(
...
+        laplacian: torch.Tensor = gaussian_pyramid[i] - img_expand
```
Comment (edgarriba) [inline kornia/geometry/transform/pyramid.py:395]: @ducha-aiki a bit confused here how we implement the difference of gaussians, over some theory i've reading too
https://github.com/kornia/kornia/blob/master/kornia/feature/responses.py#L259

https://www.cs.toronto.edu/~mangas/teaching/320/slides/CSC320L10.pdf (from slide 80)

**Review by edgarriba — COMMENTED** [review]
few more thoughs:
- should we rename (in a subsequent PR) `build_pyramid` -> `build_gaussian_pyramid`
- use separable convs to speed up things ?

File: kornia/geometry/transform/pyramid.py [hunk 11]
```diff
@@ -329,33 +331,29 @@ def build_pyramid(
...
     """
-    if not isinstance(input, torch.Tensor):
-        raise TypeError(f"Input type is not a torch.Tensor. Got {type(input)}")
-
-    if not len(input.shape) == 4:
-        raise ValueError(f"Invalid input shape, we expect BxCxHxW. Got: {input.shape}")
-
-    if not isinstance(max_level, int) or max_level < 0:
-        raise ValueError(f"Invalid max_level, it must be a positive integer. Got: {max_level}")
+    KORNIA_CHECK_IS_TENSOR(input)
```
Comment (edgarriba) [inline kornia/geometry/transform/pyramid.py:330]: Check shape already checks for a tensor

File: kornia/geometry/transform/pyramid.py (hunk 5 above)
Comment (edgarriba) [reply kornia/geometry/transform/pyramid.py:356]: For papers you can add here the bibtex and then use `:cite:` directly in the docstring

**Review by edgarriba — APPROVED** [review]

File: docs/generate_example_images.py [hunk 12]
```diff
@@ -322,7 +322,6 @@ def _get_tps_args():
         "pyrdown": ((), 1),
         "pyrup": ((), 1),
         "build_pyramid": ((3,), 1),
-        "build_laplacian_pyramid": ((3,), 1),
```
Comment (edgarriba) [inline docs/generate_example_images.py]: not adding in the end ?

File: kornia/geometry/transform/pyramid.py [hunk 13]
```diff
@@ -329,24 +327,88 @@ def build_pyramid(
...
+          The expected modes are: ``'constant'``, ``'reflect'``,
+          ``'replicate'`` or ``'circular'``.
+        align_corners: interpolation flag.
+
+    Shape:
+        - Input: :math:`(B, C, H, W)`
+        - Output :math:`[(B, C, H, W), (B, C, H/2, W/2), ...]`
+    """
+
+    KORNIA_CHECK_IS_TENSOR(input)
```
Comment (edgarriba) [inline kornia/geometry/transform/pyramid.py:386]:
```suggestion
```

File: kornia/geometry/transform/pyramid.py [hunk 14]
```diff
@@ -329,24 +327,88 @@ def build_pyramid(
...
+        input = F.pad(input, pad, "reflect")
+
+    # create gaussian pyramid
+    gaussian_pyramid: List[Tensor] = build_pyramid(input, max_level, border_type, align_corners)
+    # create empty list
+    laplacian_pyramid: List[Tensor] = []
+
+    # iterate and compute difference of adjacent layers in a gaussian pyramid
+    for i in range(max_level - 1):
+        img_expand: Tensor = pyrup(gaussian_pyramid[i + 1])
```
Comment (edgarriba) [inline kornia/geometry/transform/pyramid.py:410]:
```suggestion
        img_expand: Tensor = pyrup(gaussian_pyramid[i + 1], border_type, align_corners)
```

**Review by edgarriba — APPROVED** [review]
LGTM. Minor comments

File: kornia/geometry/transform/pyramid.py (hunk 13 above)
Comment (lafith) [reply kornia/geometry/transform/pyramid.py:386]: You want this check removed?

Comment (edgarriba) [reply kornia/geometry/transform/pyramid.py:386]: yeah, the `KORNIA_CHECK_SHAPE` already contains that

File: kornia/geometry/transform/pyramid.py [hunk 15]
```diff
@@ -5,12 +5,14 @@
 import torch.nn as nn
 import torch.nn.functional as F
 
+from kornia.core import Tensor
 from kornia.filters import filter2d, gaussian_blur2d
+from kornia.testing import KORNIA_CHECK, KORNIA_CHECK_IS_TENSOR, KORNIA_CHECK_SHAPE
```
Comment (edgarriba) [inline kornia/geometry/transform/pyramid.py:10]:
```suggestion
from kornia.testing import KORNIA_CHECK, KORNIA_CHECK_SHAPE
```

**Review by edgarriba — APPROVED** [review]

**Comment by edgarriba** [comment]
@lafith any idea why the tests are failing ?

**Pull request merged**